Added color as field with setter to Fluid class.

jabelar opened the pull request:
This is my first pull request so let me know if I'm doing something wrong. This request only affects the Forge Fluid class so I believe that means I don't need to generate a patch, let me know if I do. Also I have applied the recommended Forge formatter preferences, but let me know if my formatting is wrong.

Anyway, this is a very simple change. In the Fluid class there are several fields (viscosity, density, temperature, etc.) which are fields that have both setters and getters. However, the getColor() method is hard-coded without a field or setter, so to change it you need to extend the class and override. I don't see any reason for this. By simply making a color field with a setter, you no longer need to extend the class. So that is all I did. I created a color field and a setter method. The setter method returns the Fluid so you can chain it like the other setter methods.

The color seems to affect things like the fluid color in the universal bucket and the transparency at least and maybe the overall color affects the baked model for the fluid block. I'm also planning a future pull request to have this color affect the fog color when submerged in the fluid.

I have set the default color value to the same as before, so this should be entirely backwards compatible -- people already using getColor() will get the same value as before.

I have provided a test Mod called FluidAdditionalFieldsTest which has an ENABLE flag and resources. To test, set ENABLE to true, run client configuration and go to custom creative tab -- grab a full and empty bucket (called a "slime bag". You know the set color is working if the filled bucket shows green "slime" liquid, you can see the "slime" fluid in the world as green fluid, and if you submerge in the fluid you should see green fog. Note that the slime fluid has a low viscosity so flows slower than water.

jabelar committed:
Added color as field with setter to Fluid class. Also provides test mod called FluidAdditionalFieldsTest which has an ENABLE flag and resources. To test, set ENABLE to true, run client configuration and go to custom creative tab -- grab a full and empty bucket. You know the set color is working if the filled bucket shows green liquid, you can see the fluid in the world as green fluid, and if you submerge in the fluid you should see green fog.

Daomephsta commented:
Your test mod is 3000+ lines & includes a massive amount of superfluous stuff. Test mods should demonstrate that the feature they are testing works correctly and no more. 

This PR adds a colour field and a setter to the Fluid class, therefore the air bar stuff(L2207-L2969) is unnecessary. There are a ton of unused methods and fields everywhere as well. The slime bag is also unnecessary, use the Universal Bucket. 
In general, you need to make the test mod far more concise.

jabelar commented:
Okay, cool. As mentioned it is my first contribution so wasn't sure what might cause concern. I didn't realize that the test mod size really affects anyone since it doesn't take any real space or bandwidth. I think it is actually weirder that test mods are included in the pull request and I have to download everyone else;s (which frankly spam the console with errors and possibly conflict)...but now that I know it causes issues I can be more compact.

The actual reason though is that this is just the first pull request of about a dozen I have prepared for fluids that will use essentially the same test mod (or similar variants).  So I was going to use one file to support multiple pull requests. But I can do it the other way -- multiple small test mods instead.

jabelar committed:
As Leviathan requested I have greatly simplified the test mod, and I've made the color changes more obvious by making the assets grayscale.

jabelar commented:
Okay, Leviathan I have updated my test mod as you requested.

So to test the mod I suggest you set ENABLE to true in the FluidAdditionalFieldsTest mod, run client configuration and on the Misc Creative tab you should see a universal bucket full of green "slime". You can place the slime block in the world and confirm it is also green. Then save and quit. Go to the ModFluids class in my test mod and change the setColor() to something else like red (0xFFFF0000)  and run again and load the same world. You should now see that the slime blocks in the world are red and also in the universal bucket. You can play around with other colors (and alpha values) to see their effect. Note that the texture is already partially transparent so if you play with alpha even FF will have some transparency.

covers1624 commented:
Test mods don't exist in the mod mdk, they only exist in the forge dev environment. Only people who work directly on forge or have a special environment see them.

jabelar commented:
> Test mods don't exist in the mod mdk, they only exist in the forge dev environment. Only people who work directly on forge or have a special environment see them.

I know, that's why I was surprised that it mattered if my file was long or not (and in real terms it was only 11kB file size). I thought these test mods were something only a handful of people even see.  

But I do agree that from a testing perspective it is probably better to have the test mod as compact as possible. The only reason mine was long was because some mod functionality needs a fair bit of a full mod to test and I was planning to use one test mod to cover several pulls. I'm okay with smaller targeted ones especially since this first pull request didn't need everything.

marvin-roesch commented:
Generally speaking, test mods should both be a minimal showcase of the feature to be introduced (without using a lot of different APIs) and an example of how to use it on the code side. Furthermore, they should be exemplary Forge mods that follow best practices and promote good style.
Their purpose is not solely to demonstrate the feature's functionality to contributors, but also to serve as documentation to potential users.

jabelar commented:
> Their purpose is not solely to demonstrate the feature's functionality to contributors, but also to serve as documentation to potential users.

Okay, that seems like the best reason. I was trying to be efficient with my own time -- I wanted to use one test file for multiple pull requests. But I can see how that could be confusing in context of each separate request.

With all that out of the way -- anyone actually test my PR yet?

jabelar committed:
Made the test mod more compact as requested

jabelar commented:
Okay, made the test mod more compact as requested. Could probably even be made tighter but hopefully it is acceptable.

So to test you should enable the mod, run client, get a universal bucket of "slime" from the Misc creative tab, and you'll see that the bucket is filled with the right color. If you want you can also use the bucket to place the fluid in the world and it will also be the correct color. You can exit and edit the test mod to have a different color and run again to see the effect is applied as expected.

jabelar committed:
Expanded the imports in the test mod as requested.

jabelar commented:
Based on feedback in another PR, I have expanded the imports in my test mod.

jabelar committed:
Restoring Forge formatting as requested on another PR.

mezz commented on .gitignore:
```diff
@@ -30,3 +30,4 @@ changelog.txt
 /patches_old_bak/
 *.py
 *.7z
+/.metadata/
```
What generates this folder? If it is from an unusual utility you may want to add it to your own global git ignore instead of including it here.

jabelar replied:
Weird. I don't usually do anything to gitignore files, so not sure how that ended up here. But easy to move into the global.

mezz commented on src/main/java/net/minecraftforge/fluids/Fluid.java:
```diff
@@ -202,6 +212,19 @@ public Fluid setEmptySound(SoundEvent emptySound)
+     */
+    public Fluid setColor(int color) 
+    { 
+        this.color = color; return this; 
```
This is not a patch file, please keep one operation per line.

mezz commented on src/main/java/net/minecraftforge/fluids/Fluid.java:
```diff
@@ -202,6 +212,19 @@ public Fluid setEmptySound(SoundEvent emptySound)
+     *  of your texture, so if you want to preserve texture
+     *  then leave at default value (white).
+     */
+    public Fluid setColor(int color) 
```
It may be simpler if you take a `java.awt.Color` parameter, that way there is no question about the format.

jabelar replied:
I actually wanted to do that but the getter was an int so I felt there should be a symmetrical method. However, I'm quite happy to add a second setter method that takes Color. I'll do that.

mezz replied:
I don't think we need to keep this method, having just one with  `Color` should be good since it doesn't need the long explanation.

mezz commented on src/test/java/net/minecraftforge/debug/FluidAdditionalFieldsTest.java:
```diff
@@ -0,0 +1,81 @@
+    public static final BlockFluidBase SLIME_BLOCK = null;
+
+    @EventHandler
+    public void ev(FMLPreInitializationEvent event)
```
please use descriptive method names. `preInit` is a pretty standard name for this

jabelar replied:
okay, for event handler files, as you know, they're never called by anything so meaningful name doesn't matter, it is the parameter type that creates the meaning. However, easy change so i'll do it.

marvin-roesch replied:
While the method signature tells you *when* the method is called, it doesn't specify its *purpose*. Assume you're using one of the rendering events for a very specialized HUD element or something. You wouldn't want to call your method just `onRender` (as that's to be inferred from the signature), but instead give it a name that describes *what* you render.
Additionally, at least for me it's was easier to quickly parse the method name as that's not directly surrounded by other tokens, while the signature is embedded in between method and parameter name. Also, `preInit` is easily identifiable, while extracting that info from the clutter in `FMLPreInitializationEvent` is harder and might cause confusion with the non-Pre events at a quick glance.

mezz commented on src/test/java/net/minecraftforge/debug/FluidAdditionalFieldsTest.java:
```diff
@@ -0,0 +1,81 @@
+    }
+
+    @SubscribeEvent
+    public static void evb(final RegistryEvent.Register<Block> event)
```
method name should be descriptive, like `onBlockRegister`

mezz commented on src/test/java/net/minecraftforge/debug/FluidAdditionalFieldsTest.java:
```diff
@@ -0,0 +1,81 @@
+    }
+
+    @SubscribeEvent
+    public static void evib(final RegistryEvent.Register<Item> event)
```
method name should be descriptive, like `onItemRegister`

mezz commented on src/test/java/net/minecraftforge/debug/FluidAdditionalFieldsTest.java:
```diff
@@ -0,0 +1,81 @@
+
+    @SubscribeEvent
+    @SideOnly(Side.CLIENT)
+    public static void evm(final ModelRegistryEvent event)
```
method name should be descriptive, like `registerModels`

jabelar committed:
Changed according to suggestions. Restored the gitignore (I have no idea how it changed originally), added awt.Color option for the setter, named the test mod event handlers more explicitly names, and added a newline where there was a run-on statement in the test mod.

jabelar commented:
Okay, I addressed all of mezz's comments. Regarding the awt.Color I added an additional setter to allow people to pass it, however, I kept the field, getter and other setter as int because (a) that is the format already used (changing getter would be breaking change) and because technically the conversion would be a performance hit for the getter which is called every tick during things like fog generation.

mezz commented on src/main/java/net/minecraftforge/fluids/Fluid.java:
```diff
@@ -202,6 +214,29 @@ public Fluid setEmptySound(SoundEvent emptySound)
+     */
+    public Fluid setColor(Color color)
+    {
+        this.color = (color.getAlpha()<<24) + (color.getRed()<<16) + (color.getGreen()<<8) + color.getBlue();
```
This should be replaced with
`this.color = color.getRGB()`

jabelar replied:
Okay, done. (I had thought that maybe getRBG() didn't include alpha but it does).

mezz reviewed (changes requested):
Looks good, and yes storing the color as an `int` is the right way.

jabelar committed:
Changed the setColor() methods as per Mezz's suggestions.

liach commented on src/test/java/net/minecraftforge/debug/FluidAdditionalFieldsTest.java:
```diff
@@ -0,0 +1,84 @@
+        if (ENABLED)
+        {
+            ModelLoader.setCustomModelResourceLocation(Item.getItemFromBlock(SLIME_BLOCK), 0, new ModelResourceLocation(RES_LOC, "inventory"));
+            ModelLoader.setCustomModelResourceLocation((new ItemBlock(SLIME_BLOCK)), 0, new ModelResourceLocation(RES_LOC, "inventory"));
```
Can this line be removed?

jabelar replied:
Okay, done.

jabelar committed:
Fixed bug in test mod where there was still an int being passed to the color setter method even though it is now changed to use a Color class parameter.

jabelar committed:
Simplified the test mod as requested by liach

mezz commented:
Tested and looks good to me.

LexManos commented:
1) Why awt color? Why not just a normal Int?
2) Why do you want this? You shouldn't be modifying the color of someone else's Fluid. Why not just add this as a parameter to the constructor?

mezz commented:
awt color is more foolproof to specify and has IDE support for picking the color, the earlier version had a paragraph just describing the color format for the int.
Makes sense to put it in the constructor, but most of the fluid stuff works like a builder so I figured this way was fine.

jabelar commented:
So what is next step? I can add an alternate constructor. I can also add the alternate method for ints and a constructor for int as well. I mostly agree with mezz, but since the getter was already an int I do like having a symmetrical setter available.

jabelar committed:
As per Lex's suggestions, I have added a setter for int color and also provided alternative constructors where color can be passed as Color or as int.
